Sliding Sync: Pre-populate room data for quick filtering/sorting

MadLittleMods opened the pull request:
Pre-populate room data for quick filtering/sorting in the Sliding Sync API

Spawning from https://github.com/element-hq/synapse/pull/17450#discussion_r1697335578

This PR is acting as the Synapse version `N+1` step in the gradual migration  being tracked by https://github.com/element-hq/synapse/issues/17623

Adding two new database tables:

 - `sliding_sync_joined_rooms`: A table for storing room meta data that the local server is still participating in. The info here can be shared across all `Membership.JOIN`. Keyed on `(room_id)` and updated when the relevant room current state changes or a new event is sent in the room.
 -  `sliding_sync_membership_snapshots`: A table for storing a snapshot of room meta data at the time of the local user's membership. Keyed on `(room_id, user_id)` and only updated when a user's membership in a room changes.

Also adds background updates to populate these tables with all of the existing data.


We want to have the guarantee that if a row exists in the sliding sync tables, we are able to rely on it (accurate data). And if a row doesn't exist, we use a fallback to get the same info until the background updates fill in the rows or a new event comes in triggering it to be fully inserted. This means we need a couple extra things in place until we bump `SCHEMA_COMPAT_VERSION` and run the foreground update in the `N+2` part of the gradual migration. For context on why we can't rely on the tables without these things see [1].

 1. On start-up, block until we clear out any rows for the rooms that have had events since the max-`stream_ordering` of the `sliding_sync_joined_rooms` table (compare to max-`stream_ordering` of the `events` table). For `sliding_sync_membership_snapshots`, we can compare to the max-`stream_ordering` of `local_current_membership`
     - This accounts for when someone downgrades their Synapse version and then upgrades it again. This will ensure that we don't have any stale/out-of-date data in the `sliding_sync_joined_rooms`/`sliding_sync_membership_snapshots` tables since any new events sent in rooms would have also needed to be written to the sliding sync tables. For example a new event needs to bump `event_stream_ordering` in `sliding_sync_joined_rooms` table or some state in the room changing (like the room name). Or another example of someone's membership changing in a room affecting `sliding_sync_membership_snapshots`.
 1. Add another background update that will catch-up with any rows that were just deleted from the sliding sync tables (based on the activity in the `events`/`local_current_membership`). The rooms that need recalculating are added to the `sliding_sync_joined_rooms_to_recalculate` table.
 1. Making sure rows are fully inserted. Instead of partially inserting, we need to check if the row already exists and fully insert all data if not.

All of this extra functionality can be removed once the `SCHEMA_COMPAT_VERSION` is bumped with support for the new sliding sync tables so people can no longer downgrade (the `N+2` part of the gradual migration).


<details>
<summary><sup>[1]</sup></summary>

For `sliding_sync_joined_rooms`, since we partially insert rows as state comes in, we can't rely on the existence of the row for a given `room_id`. We can't even rely on looking at whether the background update has finished. There could still be partial rows from when someone reverted their Synapse version after the background update finished, had some state changes (or new rooms), then upgraded again and more state changes happen leaving a partial row.

For `sliding_sync_membership_snapshots`, we insert items as a whole except for the `forgotten` column ~~so we can rely on rows existing and just need to always use a fallback for the `forgotten` data. We can't use the `forgotten` column in the table for the same reasons above about `sliding_sync_joined_rooms`.~~ We could have an out-of-date membership from when someone reverted their Synapse version. (same problems as outlined for `sliding_sync_joined_rooms` above)


</details>

Discussed in an [internal meeting](https://docs.google.com/document/d/1MnuvPkaCkT_wviSQZ6YKBjiWciCBFMd-7hxyCO-OCbQ/edit#bookmark=id.dz5x6ef4mxz7)

### TODO

 - [x] Update `stream_ordering`/`bump_stamp`
 - [x] Handle remote invites
 - [x] Handle state resets
 - [x] Consider adding `sender` so we can filter `LEAVE` memberships and distinguish from kicks.
     - [x] We should add it to be able to tell leaves from kicks 
 - [x] Consider adding `tombstone` state to help address https://github.com/element-hq/synapse/issues/17540
     - [x] We should add it `tombstone_successor_room_id`
 - [x] Consider adding `forgotten` status to avoid extra lookup/table-join on `room_memberships`
    - [x] We should add it
 - [x] Background update to fill in values for all joined rooms and non-join membership
 - [x] Clean-up tables when room is deleted
 - [x] Make sure tables are useful to our use case
    - First explored in https://github.com/element-hq/synapse/compare/erikj/ss_use_new_tables
    - Also explored in https://github.com/element-hq/synapse/commit/76b5a576eb363496315dfd39510cad7d02b0fc73
    - -> New tables utilized in https://github.com/element-hq/synapse/pull/17630
 - [x] Plan for how can we use this with a fallback
     - See plan discussed above in main area of the issue description
     - Discussed in an [internal meeting](https://docs.google.com/document/d/1MnuvPkaCkT_wviSQZ6YKBjiWciCBFMd-7hxyCO-OCbQ/edit#bookmark=id.dz5x6ef4mxz7)
 - [x] Plan for how we can rely on this new table without a fallback
     - Synapse version `N+1`: (this PR) Bump `SCHEMA_VERSION` to `87`. Add new tables and background update to backfill all rows. Since this is a new table, we don't have to add any `NOT VALID` constraints and validate them when the background update completes. Read from new tables with a fallback in cases where the rows aren't filled in yet.
     - Synapse version `N+2`: Bump `SCHEMA_VERSION` to `88` and bump `SCHEMA_COMPAT_VERSION` to `87` because we don't want people to downgrade and miss writes while they are on an older version. Add a foreground update to finish off the backfill so we can read from new tables without the fallback. Application code can now rely on the new tables being populated.
     - Discussed in an [internal meeting](https://docs.google.com/document/d/1MnuvPkaCkT_wviSQZ6YKBjiWciCBFMd-7hxyCO-OCbQ/edit#bookmark=id.hh7shg4cxdhj)




### Dev notes

```
SYNAPSE_TEST_LOG_LEVEL=INFO poetry run trial tests.storage.test_events.SlidingSyncPrePopulatedTablesTestCase

SYNAPSE_POSTGRES=1 SYNAPSE_POSTGRES_USER=postgres SYNAPSE_TEST_LOG_LEVEL=INFO poetry run trial tests.storage.test_events.SlidingSyncPrePopulatedTablesTestCase
```

```
SYNAPSE_TEST_LOG_LEVEL=INFO poetry run trial tests.handlers.test_sliding_sync.FilterRoomsTestCase
```

Reference:

 - [Development docs on background updates and worked examples of gradual migrations
](https://github.com/element-hq/synapse/blob/1dfa59b238cee0dc62163588cc9481896c288979/docs/development/database_schema.md#background-updates)
 - A real example of a gradual migration: https://github.com/matrix-org/synapse/pull/15649#discussion_r1213779514
 - Adding `rooms.creator` field that needed a background update to backfill data, https://github.com/matrix-org/synapse/pull/10697
 - Adding `rooms.room_version` that needed a background update to backfill data, https://github.com/matrix-org/synapse/pull/6729
 - Adding `room_stats_state.room_type` that needed a background update to backfill data, https://github.com/matrix-org/synapse/pull/13031
 - Tables from MSC2716: `insertion_events`, `insertion_event_edges`, `insertion_event_extremities`, `batch_events`
 - `current_state_events` updated in `synapse/storage/databases/main/events.py`

---

```
persist_event (adds to queue)
_persist_event_batch
_persist_events_and_state_updates (assigns `stream_ordering` to events)
_persist_events_txn
	_store_event_txn
        _update_metadata_tables_txn
            _store_room_members_txn
	_update_current_state_txn
```

---

> Concatenated Indexes [...] (also known as multi-column, composite or combined index)
>
> [...] key consists of multiple columns.
> 
> We can take advantage of the fact that the first index column is always usable for searching
>
> *-- https://use-the-index-luke.com/sql/where-clause/the-equals-operator/concatenated-keys*

---

Dealing with `portdb` (`synapse/_scripts/synapse_port_db.py`), https://github.com/element-hq/synapse/pull/17512#discussion_r1725998219

---

<details>
<summary>SQL queries:</summary>

Both of these are equivalent and work in SQLite and Postgres

Options 1:
```sql
WITH data_table (room_id, user_id, membership_event_id, membership, event_stream_ordering, {", ".join(insert_keys)}) AS (
    VALUES (
        ?, ?, ?,
        (SELECT membership FROM room_memberships WHERE event_id = ?),
        (SELECT stream_ordering FROM events WHERE event_id = ?),
        {", ".join("?" for _ in insert_values)}
    )
)
INSERT INTO sliding_sync_non_join_memberships
    (room_id, user_id, membership_event_id, membership, event_stream_ordering, {", ".join(insert_keys)})
SELECT * FROM data_table
WHERE membership != ?
ON CONFLICT (room_id, user_id)
DO UPDATE SET
    membership_event_id = EXCLUDED.membership_event_id,
    membership = EXCLUDED.membership,
    event_stream_ordering = EXCLUDED.event_stream_ordering,
    {", ".join(f"{key} = EXCLUDED.{key}" for key in insert_keys)}
```

Option 2:
```sql
INSERT INTO sliding_sync_non_join_memberships
    (room_id, user_id, membership_event_id, membership, event_stream_ordering, {", ".join(insert_keys)})
SELECT 
    column1 as room_id,
    column2 as user_id,
    column3 as membership_event_id,
    column4 as membership,
    column5 as event_stream_ordering,
    {", ".join("column" + str(i) for i in range(6, 6 + len(insert_keys)))}
FROM (
    VALUES (
        ?, ?, ?,
        (SELECT membership FROM room_memberships WHERE event_id = ?),
        (SELECT stream_ordering FROM events WHERE event_id = ?),
        {", ".join("?" for _ in insert_values)}
    )
) as v
WHERE membership != ?
ON CONFLICT (room_id, user_id)
DO UPDATE SET
    membership_event_id = EXCLUDED.membership_event_id,
    membership = EXCLUDED.membership,
    event_stream_ordering = EXCLUDED.event_stream_ordering,
    {", ".join(f"{key} = EXCLUDED.{key}" for key in insert_keys)}
```

If we don't need the `membership` condition, we could use:

```sql
INSERT INTO sliding_sync_non_join_memberships
    (room_id, membership_event_id, user_id, membership, event_stream_ordering, {", ".join(insert_keys)})
VALUES (
    ?, ?, ?,
    (SELECT membership FROM room_memberships WHERE event_id = ?),
    (SELECT stream_ordering FROM events WHERE event_id = ?),
    {", ".join("?" for _ in insert_values)}
)
ON CONFLICT (room_id, user_id)
DO UPDATE SET
    membership_event_id = EXCLUDED.membership_event_id,
    membership = EXCLUDED.membership,
    event_stream_ordering = EXCLUDED.event_stream_ordering,
    {", ".join(f"{key} = EXCLUDED.{key}" for key in insert_keys)}
```

</details>

### Pull Request Checklist

<!-- Please read https://element-hq.github.io/synapse/latest/development/contributing_guide.html before submitting your pull request -->

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

MadLittleMods committed:
Start thinking about schemas

MadLittleMods committed:
Add changelog

MadLittleMods committed:
Use foreign keys

MadLittleMods committed:
Merge branch 'develop' into madlittlemods/sliding-sync-pre-populate-room-meta-data

MadLittleMods committed:
Start of updating `sliding_sync_joined_rooms`

MadLittleMods committed:
Fill in `sliding_sync_non_join_memberships` when current state changes

MadLittleMods committed:
Special treatment for boolean columns

See https://github.com/element-hq/synapse/blob/1dfa59b238cee0dc62163588cc9481896c288979/docs/development/database_schema.md#boolean-columns

MadLittleMods committed:
Test is running

MadLittleMods committed:
Server left room test

MadLittleMods committed:
Change to updating the latest membership in the room

MadLittleMods committed:
Closer to right

MadLittleMods committed:
Fix comparison and insert

MadLittleMods committed:
Better test assertions

MadLittleMods committed:
Test non-joins

MadLittleMods committed:
Add more tests

MadLittleMods committed:
Handle to_delete

MadLittleMods committed:
Handle server left room

MadLittleMods committed:
Fix some lints

MadLittleMods commented on synapse/storage/schema/main/delta/87/01_sliding_sync_memberships.sql:
```diff
@@ -0,0 +1,57 @@
+-- the membership event itself as the `bump_stamp`.
+CREATE TABLE IF NOT EXISTS sliding_sync_non_join_memberships(
+    room_id TEXT NOT NULL REFERENCES rooms(room_id),
+    user_id TEXT NOT NULL,
+    membership_event_id TEXT NOT NULL REFERENCES events(event_id),
+    membership TEXT NOT NULL,
```
Just adding all of these fields here because I'm not sure what we're most likely to look things up by and it's unclear if we would prefer to join on `local_current_membership` or not

MadLittleMods committed:
Fill in `stream_ordering`/`bump_stamp` when we add current state to the joined rooms table

MadLittleMods committed:
Fill in `stream_ordering`/`bump_stamp` for any event being persisted

MadLittleMods committed:
Need to fix upsert

MadLittleMods committed:
Fix bumping when events are persisted out of order

MadLittleMods committed:
Refactor to `sliding_sync_membership_snapshots`

MadLittleMods committed:
Update descriptions

MadLittleMods committed:
Fix lints

MadLittleMods committed:
Fill in for remote invites (out of band, outlier membership)

MadLittleMods committed:
Fix events from rooms we're not joined to affecting the joined room stream ordering

MadLittleMods committed:
User ID is not unique because user is joined to many rooms

MadLittleMods committed:
Add tests for remote invites

MadLittleMods committed:
Add test for remote invite rejected/retracted

MadLittleMods committed:
Add test to make sure snapshot evolves with membership

MadLittleMods committed:
Adjust test description

MadLittleMods committed:
Fix lints

MadLittleMods committed:
Fix federating backfill test

MadLittleMods committed:
Merge branch 'develop' into madlittlemods/sliding-sync-pre-populate-room-meta-data

MadLittleMods committed:
Fix rejecting invite when no_longer_in_room (and other non-join transitions)

MadLittleMods committed:
Clarify when/why we upsert

MadLittleMods committed:
Update fixme comment

MadLittleMods committed:
Fix boolean schema for Postgres

MadLittleMods commented on synapse/storage/databases/main/events.py:
```diff
@@ -1260,6 +1468,135 @@ def _update_current_state_txn(
+                # `_update_sliding_sync_tables_with_new_persisted_events_txn()` is run
+                # after this function to update it to the correct latest value. This is
+                # just to account for things changing in the future.
+                next(iter(to_insert.values())),
```
When `to_insert` is empty, this raises a `StopIteration` exception/error that is completely swallowed by the transaction and silenty errors (never logs anything went wrong).

Why does `raise StopIteration()` not get caught by the [database trasnsaction `except Exception` handler](https://github.com/element-hq/synapse/blob/a308d99f30d7e660115e355c54c37ac149cdbe53/synapse/storage/database.py#L938-L941) but `raise AssertionError("fake error")` does? They are both `Exception`

MadLittleMods replied:
Seems related to https://peps.python.org/pep-0479/ but that makes it seem like with that change (since Python 3.5), `StopIteration` shouldn't be silently ignored like it is now.

I'm guessing this is a bad interaction between the `with` context managers at various points and `make_deferred_yieldable(...)` swallowing it.

There have been more recent related issues around this like https://github.com/python/cpython/issues/112182

MadLittleMods replied:
Created https://github.com/element-hq/synapse/issues/17597 to track this

MadLittleMods committed:
Fix running into `StopIteration`

More context about how/why `StopIteration` was being ignored silently
which made this problem harder to debug. See
https://github.com/element-hq/synapse/pull/17512#discussion_r1715954505

MadLittleMods committed:
More realistic remote room forgotten test

MadLittleMods committed:
No invites needed

MadLittleMods committed:
Clean up tables when a room is purged/deleted

MadLittleMods committed:
Fix lints

MadLittleMods commented on tests/rest/client/sliding_sync/test_rooms_meta.py:
```diff
@@ -691,7 +708,22 @@ def test_rooms_bump_stamp_backfill(self) -> None:
+                partial_state=False,
+            )
+        )
+        self.get_success(self.persistence.persist_event(join_event, join_context))
```
I had to update this test (`test_rooms_bump_stamp_backfill`) and the one below (`test_join_locally_forgotten_room`) to use a more realistic remote room join (one where we actually have `current_state_events` after joining). Otherwise, this kinda breaks our assumptions in Synapse. This was fine previously but I've added an assert to make sure we have `current_state_events` unless we are `no_longer_in_room`.

MadLittleMods committed:
Add background update for `sliding_sync_joined_rooms`

MadLittleMods committed:
Add more tests

MadLittleMods committed:
Fix lints

MadLittleMods committed:
Work on background update for `sliding_sync_membership_snapshots`

MadLittleMods committed:
Use helper

MadLittleMods committed:
Move background updates to `StateBackgroundUpdateStore`

So we can access `_get_state_groups_from_groups_txn(...)`

MadLittleMods committed:
Add historical case to background update

MadLittleMods committed:
Add background update test for `sliding_sync_membership_snapshots`

MadLittleMods committed:
Add more tests

MadLittleMods committed:
Finish off background update tests

MadLittleMods committed:
Fix lints

MadLittleMods committed:
Move function next to other helpers

MadLittleMods committed:
Fill in docstrings

MadLittleMods committed:
Update comment

MadLittleMods commented on changelog.d/17512.misc:
```diff
@@ -0,0 +1 @@
+Pre-populate room data used in experimental [MSC3575](https://github.com/matrix-org/matrix-spec-proposals/pull/3575) Sliding Sync `/sync` endpoint for quick filtering/sorting.
```
@erikjohnston you mentioned wanting to have a look at this to start getting familiar and light review.

It's at the point where all of the pieces are in place to update the tables and background updates to backfill all of the data.

There are a few items left in the TODO section of the PR description. The main one is that I haven't had a chance to actually try out whether these schema's actually make sense in our use case yet.

erikjohnston commented on synapse/storage/databases/main/events.py:
```diff
@@ -1260,6 +1411,100 @@ def _update_current_state_txn(
+                    state_key[0] in SLIDING_SYNC_DEFAULT_BUMP_EVENT_TYPES
+                    and state_key[1] == ""
+                ):
+                    bump_event_id = event_id
```
Why are we doing these here? I think we should just leave it to `_update_sliding_sync_tables_with_new_persisted_events_txn`?

MadLittleMods replied:
To make the insert more "complete" but I can skip.

It's all done in the same transaction so pretty much the same thing anyway :thumbsup:

MadLittleMods replied:
Updated :white_check_mark:

erikjohnston commented on synapse/storage/databases/main/events.py:
```diff
@@ -1303,12 +1548,276 @@ def _update_current_state_txn(
+                    f"Unexpected event (we should not be fetching extra events): ({event_type}, {state_key})"
+                )
+
+        return sliding_sync_insert_map
```
Given these need fetch events, and so we need to handle redactions, I think we should do this logic outside of the persist events transaction (maybe as part of calculating the current state deltas?).

I think that might also make some of this logic a bit clearer as well, as it divorces calculating what to add with actually persisting it.

The other benefit is ensuring this transaction remains reasonably fast, this is important as (for Reasons) a slow persist event transaction can cause events persisted in other rooms to be delayed being sent down sync

erikjohnston replied:
Something around here maybe: https://github.com/element-hq/synapse/blob/f162c92f2a1f8cf41b5f7211cb31cf8ae49b20e7/synapse/storage/controllers/persist_events.py#L608-L629

erikjohnston replied:
I'm guessing the returned struct could look something like:

```python
class SlidingSyncTableChanges:
    # room_id -> dict to upsert into `sliding_sync_joined_rooms`
    joined_room_updates: Dict[str, JsonDict]
    # (room_id, user_id) -> dict to upsert into sliding_sync_membership_snapshots
    membership_updates: Dict[Tuple[str, str], JsonDict]

    # room_ids to delete from `sliding_sync_joined_rooms`
    to_delete_from_joined_rooms: StrCollection
```

MadLittleMods replied:
I've updated the normal update loops to do the pre-processing outside of the transaction which solves the redaction case there.

I'm not sure how to do the same for the background update though since the main store methods don't seem to be available. And I don't see an existing function which fetches events taking redactions into account (something we could make though).

In any case, probably just need a quick chat with you for a way forward there.

erikjohnston replied:
> I'm not sure how to do the same for the background updates though since the main store methods don't seem to be available. And I don't see an existing function which fetches events taking redactions into account (something we could make though).

I'm fairly sure that `get_events` is available for background updates, though I can imagine `mypy` might get confused given they're mixin classes.


> Additionally, even if we were able to use the main store methods, since it's not being all done in one transaction, for `_sliding_sync_joined_rooms_backfill`, I can't guarantee that I'm inserting the actual latest data or racing with the normal update loop. I would need to lock the table for that row or something.

I think you can probably record the latest current state delta stream ID before you calculate the entry for a room, and then in the transaction to persist the new data check for relevant changes in `current_state_delta_stream` since the stream ID to see whether your calculated entries are now out of date. For the event stream ordering and bump stamp I think you just need to assert you're not overwriting an existing value (the existing value will be up to date if it exists), the same goes for the state too though you need to be careful of state resets (hence looking at the current_state_delta_stream table again).

MadLittleMods replied:
Talked with @erikjohnston a bit more. It's a bit tricky to put the background updates in the "right" place since we need access to state at a given event (`state` store) but we're updating the table in the `main` store. We haven't done a background update like this before.

The suggestion was to move to the `main` store (as we're updating the table in the `main` store), and then get the `state` storage via `hs.get_storage_controllers()`. Alternatively, add a function in e.g. `PersistEventsController` or whatever to add an `async fn` that drives the update.

I've updated things to go with the `main` store route and using `self.hs.get_storage_controllers().state` to get at the state necessary :white_check_mark: 

---

I was also able to use the suggestion to check the `current_state_delta_stream` to see whether the `current_state_events` pulled out are up-to-date or not. Currently, I just throw an exception if we see that they are out of date (saving our place just before exiting) and have the background update retry logic kick in.

Thanks for the help unblocking here @erikjohnston :slightly_smiling_face:

erikjohnston commented on synapse/storage/databases/main/events.py:
```diff
@@ -1303,12 +1548,276 @@ def _update_current_state_txn(
+                    WHEN event_stream_ordering IS NULL OR event_stream_ordering < ?
+                        THEN ?
+                    ELSE event_stream_ordering
+                END,
```
We actually know these things will be true, as stream ordering always increases

MadLittleMods replied:
I understand that `stream_ordering` is unique and is incremented for each event but are we always sure that events are persisted in that same order? Given that we also process backfilled events here, seems like that isn't a guarantee.

erikjohnston reviewed (commented):
This looks broadly on-track. The only thing I'm wondering really is if we should be doing some of the calculations outside the persist event transaction. I think the logic is sound though

MadLittleMods commented on synapse/storage/databases/main/events.py:
```diff
@@ -1158,12 +1188,127 @@ def _update_current_state_txn(
+            membership_event_id_to_user_id_map: Dict[str, str] = {}
+            for state_key, event_id in to_insert.items():
+                if state_key[0] == EventTypes.Member and self.is_mine_id(state_key[1]):
+                    membership_event_id_to_user_id_map[event_id] = state_key[1]
```
I was initially tracking only [non-joins](https://github.com/element-hq/synapse/pull/17512/commits/0af3b4822c5a254f81458017764d7f35aa3cc9b9) but then realized it might be easier just to track any membership. This way we don't need to delete the row from the table when the person joins the room (less upkeep) and we can probably use this table to replace `get_rooms_for_local_user_where_membership_is(...)` much better and with a whole lot less table joins if we have the complete picture. We can query `sliding_sync_membership_snapshots` and then join to `sliding_sync_joined_rooms` for any `JOIN` rooms.

We can even `ORDER BY event_stream_ordering` and start grabbing less out of the database to only fill up what we need for the given `range` requested.

MadLittleMods committed:
Merge branch 'develop' into madlittlemods/sliding-sync-pre-populate-room-meta-data

MadLittleMods commented on synapse/storage/databases/main/events.py:
```diff
@@ -1158,12 +1188,127 @@ def _update_current_state_txn(
+                    for event_type, state_key in to_delete
+                    if event_type == EventTypes.Member and self.is_mine_id(state_key)
+                ),
+            )
```
If other bits of state reset, I assume that appears as a `to_insert` though? In which case, I think we're covered

MadLittleMods replied:
Oh, I think I know what you mean.

For `sliding_sync_joined_rooms`, we handle any relevant bits of state resetting (see `to_delete` below).

For the `sliding_sync_membership_snapshots` though, I don't think we should be updating the snapshots once set. And we already cover bits of state being reset in the same batch as some snapshots are being added (see `to_delete` below)

MadLittleMods committed:
Prefer `_update_sliding_sync_tables_with_new_persisted_events_txn(...)` to do the right thing

See https://github.com/element-hq/synapse/pull/17512#discussion_r1719992152

MadLittleMods committed:
Add test to handle state reset in the meta data

MadLittleMods committed:
Test state reset on membership

MadLittleMods committed:
Start of moving logic outside of the transaction (pre-process)

MadLittleMods committed:
Closer types

MadLittleMods committed:
Use dicts

MadLittleMods committed:
Use `TypedDict`

MadLittleMods committed:
Move pre-processing completely outside transaction

MadLittleMods committed:
Add sanity checks and fix wrong variable usage

MadLittleMods committed:
Merge branch 'develop' into madlittlemods/sliding-sync-pre-populate-room-meta-data

MadLittleMods commented on synapse/storage/controllers/persist_events.py:
```diff
@@ -751,6 +775,261 @@ async def _calculate_new_forward_extremities_and_state_delta(
+
+            if membership_infos_to_insert_membership_snapshots:
+                current_state_ids_map: MutableStateMap = dict(
+                    await self.main_store.get_partial_filtered_current_state_ids(
```
Because we're doing the pre-processing outside of the transaction now. There is no guarantee that we're getting the correct `current_state_events` here.

We probably have to update `_calculate_new_forward_extremities_and_state_delta(...)` to return the `current_state` that it calculates to generate the `DeltaState`

MadLittleMods replied:
I guess this might be fine since each processing task is done one-by-one.

We are already layering on top `to_insert`/`to_delete` but that wasn't the concern here.

erikjohnston replied:
Worth noting that nothing else can actually be inserting new events or current state into the room at this point. All this code id being run as part of the insert event queue

MadLittleMods committed:
Attempt getting real events in backgroun update (needs work)

MadLittleMods committed:
Go back to simpler fetching senders

erikjohnston commented on synapse/storage/databases/main/events.py:
```diff
@@ -1260,6 +1366,75 @@ def _update_current_state_txn(
+                        VALUES (
+                            ?,
+                            (SELECT stream_ordering FROM events WHERE event_id = ?),
+                            {", ".join("?" for _ in insert_values)}
```
Can we not just use the stream ordering from the latest event we're persisting here? I think `stream_id` might actually be it?

MadLittleMods replied:
ohh, I hadn't noticed `stream_id` here (the events themselves aren't available in this function). Updated to use it :white_check_mark:

MadLittleMods replied:
I had tried this out in https://github.com/element-hq/synapse/pull/17512/commits/772c501bb6c1d1fb2a2fc85025495a6268bdd75a but the [CI actually starts failing](https://github.com/element-hq/synapse/actions/runs/10497666048/job/29081101793?pr=17512#step:7:13839) because of this [test about partial joins](https://github.com/matrix-org/complement/blob/6e4426a9e63233f9821a4d2382bfed145244183f/tests/msc3902/federation_room_join_partial_state_test.go#L3625-L3668) because the `stream_id` we are trying to use doesn't exist.

To run the test locally:
```
POSTGRES=1 COMPLEMENT_ALWAYS_PRINT_SERVER_LOGS=1 COMPLEMENT_DIR=../complement ./scripts-dev/complement.sh -run TestPartialStateJoin/Leave_during_resync/is_seen_after_the_resync  -p 1
...
psycopg2.errors.ForeignKeyViolation: insert or update on table "sliding_sync_joined_rooms" violates foreign key constraint "sliding_sync_joined_rooms_event_stream_ordering_fkey"
synapse_main | DETAIL:  Key (event_stream_ordering)=(4) is not present in table "events".
```


### Root cause

We seem to have some confused usage or at-least disparate usage of what `stream_id` should be.

Here we call `_update_current_state_txn(...)` with the [`min_stream_order`](https://github.com/element-hq/synapse/blob/92b38c1afd1acd1b88b3ae313f05bcf1bec0e849/synapse/storage/databases/main/events.py#L593-L595) of the batch of events we're persisting (seems correct :white_check_mark:).

But we also seem to call it with the [next upcoming stream_ordering](https://github.com/element-hq/synapse/blob/92b38c1afd1acd1b88b3ae313f05bcf1bec0e849/synapse/storage/databases/main/events.py#L1137-L1144). This code path is exercised when we try to [re-sync the state of a partial stated room](https://github.com/element-hq/synapse/blob/92b38c1afd1acd1b88b3ae313f05bcf1bec0e849/synapse/handlers/federation.py#L1933). Which makes sense why we see this failure with a partial join room test. I see how this could be "correct" but it means we can't rely on `stream_id` because we expect to point to an existing event.

Maybe there is a better way to handle the partial state resync case but it requires more thought and a separate PR.

MadLittleMods replied:
The `stream_id` is necessary for the `current_state_delta_stream` table and to bust caches so I don't think we can make it optional and supply `None`.

We could figure out *some* `stream_ordering` to use in `_calculate_sliding_sync_table_changes(...)` and accomplish this setup though.

erikjohnston replied:
Yeah, I was thinking of having both the `stream_id` and an optional `event_stream_ordering` (or whatever) to use when inserting? (Not totally sure how feasible that is)

MadLittleMods replied:
Here is what I ended up with and simplified to just the `simple_upsert_txn()` -> https://github.com/element-hq/synapse/pull/17512/files/6edc4c78ce5c7f4abbc27e68df8cf2038520d625..4b866c4fcad9238d8f876be4d53b04a3e43da2c3 (this is a link to view 3 commits that make up the change)

I decided against continuing the optimization of just doing an `UPDATE` if we only have things `to_delete`. This gives us less paths to reason about (simple).

MadLittleMods replied:
More conversation in https://github.com/element-hq/synapse/pull/17732#discussion_r1794398423

erikjohnston commented on synapse/storage/databases/main/events.py:
```diff
@@ -1296,19 +1471,347 @@ def _update_current_state_txn(
+                    {("," + ", ".join(insert_keys)) if insert_keys else ""})
+                VALUES (
+                    ?, ?, ?,
+                    (SELECT membership FROM room_memberships WHERE event_id = ?),
+                    (SELECT stream_ordering FROM events WHERE event_id = ?)
```
It feels like we would have already pulled out these values when figuring out what to insert?

These are mainly just suggestions that might make the queries simpler

MadLittleMods replied:
I tried that https://github.com/element-hq/synapse/pull/17512/commits/357132db1d4216a1e74d7234f443c9a2b38c1687 but the `event.internal_metadata.stream_ordering` was actually `None` and my assumption was wrong. Or maybe one of the tests is using a slightly less real setup where it comes through as `None` at that point

MadLittleMods replied:
In `_persist_event_batch()`, `_calculate_sliding_sync_table_changes()` is called before `_persist_events_and_state_updates()`  so that's why we're seeing all of the `event.internal_metadata.stream_ordering` as `None`

MadLittleMods replied:
I could move `_calculate_sliding_sync_table_changes(...)` inside `_persist_events_and_state_updates(...)` after we give the events their `stream_ordering` but before `_persist_events_txn(...)`. Seems viable, will give it a try :runner:

MadLittleMods replied:
Updated and works :white_check_mark:

MadLittleMods committed:
Use full event version after solving the circular import issues

MadLittleMods committed:
Move back to the main store

MadLittleMods committed:
Get full events for `_sliding_sync_joined_rooms_backfill`

MadLittleMods committed:
No need to update `event_stream_ordering`/`bump_stamp` `ON CONFLICT`

MadLittleMods committed:
Use available `stream_id`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1725310035

MadLittleMods committed:
Move `_calculate_sliding_sync_table_changes(...)` after we assign `stream_ordering` to events

See https://github.com/element-hq/synapse/pull/17512#discussion_r1725728637

MadLittleMods committed:
Pre-populate `membership` and `membership_event_stream_ordering`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1725311745

MadLittleMods committed:
Add `sender` column so we can tell leaves from kicks

MadLittleMods committed:
Move away from `stream_id`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1725806543

MadLittleMods committed:
Add `tombstone_successor_room_id` column

MadLittleMods committed:
Update tests

MadLittleMods committed:
Add tombstone to tests

MadLittleMods committed:
Better handle none case

MadLittleMods committed:
Log which room is strange

MadLittleMods commented on synapse/storage/databases/main/events_bg_updates.py:
```diff
@@ -1516,3 +1541,491 @@ def _populate_txn(txn: LoggingTransaction) -> bool:
+            bump_stamp_event_pos_results = await self.get_last_event_pos_in_room(
+                room_id, event_types=SLIDING_SYNC_DEFAULT_BUMP_EVENT_TYPES
+            )
+            assert bump_stamp_event_pos_results, (
+                f"We should not be seeing `None` here because the room ({room_id}) should at-least have a create event "
+                + "(unless `SLIDING_SYNC_DEFAULT_BUMP_EVENT_TYPES` no longer includes the room create event)"
+            )
```
It looks like the [`portdb` CI job](https://github.com/element-hq/synapse/actions/runs/10499584913/job/29086695919?pr=17512) is failing because our test database (`.ci/test_db.db`) has a room (`!0:localhost:42455`) that has `current_state_events` (which means it's picked up in this background update) but the `m.room.create` event is an `outlier` so we can't find any bump event.

The only non-outlier is the join membership at the end. This room seems malformed (if we have `current_state_events`, feels like those events should be de-outliered) and I don't think we should try to handle this scenario.

Keep in mind, I'm looking at the database before all of the background updates have run on it:
```sh
$ sqlite3 .ci/test_db.db

> SELECT * FROM current_state_events WHERE room_id = '!0:localhost:42455';
$0:localhost:42455|!0:localhost:42455|m.room.create||
$2:localhost:42455|!0:localhost:42455|m.room.join_rules||
$1:localhost:42455|!0:localhost:42455|m.room.member|@__ANON__-0:localhost:42455|join
$1570043341273LsXNn:localhost:8800|!0:localhost:42455|m.room.member|@anon-20191002_181700-617:localhost:8800|join

> SELECT type, outlier, event_id FROM events WHERE room_id = '!0:localhost:42455';
m.room.create|1|$0:localhost:42455
m.room.member|1|$1:localhost:42455
m.room.join_rules|1|$2:localhost:42455
m.room.member|0|$1570043341273LsXNn:localhost:8800
```

erikjohnston replied:
Yeah, this is a legit scenario. When you join a federated room you won't get the full history, and so the `create` will be an `outlier`.

MadLittleMods replied:
Ahh, this makes sense :+1:

MadLittleMods committed:
Add missing boolean column to portdb script

MadLittleMods committed:
Add `forgotten` column

MadLittleMods committed:
Fill out docstring

MadLittleMods committed:
More docstring

MadLittleMods commented on synapse/storage/schema/main/delta/87/01_sliding_sync_memberships.sql:
```diff
@@ -0,0 +1,134 @@
+    sender TEXT NOT NULL,
+    membership_event_id TEXT NOT NULL REFERENCES events(event_id),
+    membership TEXT NOT NULL,
+    -- This is an integer just to match `room_memberships` and also means we don't need
+    -- to do any casting.
+    forgotten INTEGER DEFAULT 0 NOT NULL,
```
Is it okay that this is an `INTEGER` to match `room_memberships`? Ideally this would be `BOOLEAN`.

Matching `room_memberships` saves us that familiarity and we can avoid tricky casting when we do a subquery like `(SELECT forgotten FROM room_memberships WHERE room_id = ? AND user_id = ?)`

erikjohnston replied:
Yeah, I think it makes sense to be consistent.

MadLittleMods committed:
Fix sub-query selecting multiple rows

erikjohnston commented on synapse/storage/databases/main/events.py:
```diff
@@ -1296,19 +1735,371 @@ def _update_current_state_txn(
+
+    @classmethod
+    def _get_sliding_sync_insert_values_from_stripped_state_txn(
+        cls, txn: LoggingTransaction, unsigned_stripped_state_events: Any
```
This doesn't need to be in a transcation?

MadLittleMods replied:
Good catch :+1:

erikjohnston commented on tests/storage/test_events.py:
```diff
@@ -481,3 +491,4120 @@ def test_room_remote_user_cache_invalidated(self) -> None:
+            set(sliding_sync_membership_snapshots_results.keys()),
+            set(),
+            exact=True,
+        )
```
I think we should move these tests to a dedicated file.

erikjohnston reviewed (commented):
Looks good, just I think some cleanups that will make things easier to follow

MadLittleMods commented on synapse/storage/databases/main/events.py:
```diff
@@ -226,6 +310,14 @@ async def _persist_events_and_state_updates(
                 event.internal_metadata.instance_name = self._instance_name
 
+            sliding_sync_table_changes = None
+            if state_delta_for_room is not None:
```
We handle this in `_store_room_members_txn`, find this comment:
```
# Handle updating the `sliding_sync_membership_snapshots` table
# (out-of-band membership events only)
```

MadLittleMods committed:
Check `events_and_context` for state events

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726798803

MadLittleMods committed:
Rename `insert_key`/`insert_value`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726805894

MadLittleMods committed:
Prefer `simple_update_txn`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726808112

MadLittleMods committed:
Prefer `simple_delete_many_txn`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726819380

MadLittleMods commented on synapse/storage/databases/main/events.py:
```diff
@@ -590,10 +905,21 @@ def _persist_events_txn(
+                sliding_sync_table_changes,
             )
 
+        self._update_sliding_sync_tables_with_new_persisted_events_txn(
+            txn, events_and_contexts
+        )
```
`use_negative_stream_ordering` isn't available here. We do check for this inside `_update_sliding_sync_tables_with_new_persisted_events_txn()` though by checking if the first event has a negative `stream_ordering` (and assume the whole batch is backfilled).

erikjohnston replied:
Ah, must have misread the function signatures. I thought at some point we passed in `backfilled: bool` to the transaction, but looks its been cleared up. No biggie though, just thought if there was a flag may as well use it.

MadLittleMods committed:
Simplify `_update_sliding_sync_tables_with_new_persisted_events_txn()`

See
https://github.com/element-hq/synapse/pull/17512#discussion_r1719997640
https://github.com/element-hq/synapse/pull/17512#discussion_r1726828894
https://github.com/element-hq/synapse/pull/17512#discussion_r1726836440

MadLittleMods committed:
"backfill" -> "bg_update"

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726837698

MadLittleMods committed:
Move away from backfill language

MadLittleMods committed:
No need for transaction

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726844107

MadLittleMods committed:
Move tests to dedicated file

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726849798

MadLittleMods committed:
Allow for no `bump_stamp` (fix `portdb` CI job)

See https://github.com/element-hq/synapse/pull/17512#discussion_r1725998219

MadLittleMods committed:
Derive best effort `stream_ordering` outside of the transaction

See https://github.com/element-hq/synapse/pull/17512#discussion_r1727995882

MadLittleMods committed:
Use `simple_upsert_txn` to update `sliding_sync_joined_rooms`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726817206

MadLittleMods committed:
Simplify what we need to think about to grab the best effort value

MadLittleMods committed:
Correct comment

MadLittleMods committed:
Use `simple_upsert_many_txn` for `sliding_sync_membership_snapshots`

See https://github.com/element-hq/synapse/pull/17512#discussion_r1726820006

MadLittleMods commented on synapse/storage/databases/main/events_bg_updates.py:
```diff
@@ -1516,3 +1541,500 @@ def _populate_txn(txn: LoggingTransaction) -> bool:
+                    ON CONFLICT (room_id)
+                    DO UPDATE SET
+                        {", ".join(f"{key} = EXCLUDED.{key}" for key in insert_keys)}
+                    """
```
Good call, this has evolved to not have any of those sub-queries anymore :+1:

MadLittleMods replied:
I've updated them to use `simple_upsert_txn`. I can't use the `simple_upsert_many_txn` variant for `sliding_sync_membership_snapshots` because each one has its' own `insert_map`. And I can't use it for `sliding_sync_membership_snapshots` because we do them one by one to see if our current state is out of date by the time we insert and retry if necessary.

MadLittleMods commented on synapse/storage/databases/main/events.py:
```diff
@@ -1296,19 +1735,371 @@ def _update_current_state_txn(
+                    + list(insert_values)
+                    for membership_info in sliding_sync_table_changes.to_insert_membership_snapshots
+                ],
+            )
```
Good call, this has evolved to not have any of those sub-queries anymore :+1:

MadLittleMods committed:
Fix partial-stated room re-syncing state but nothing has changed

Fixes failing test in CI: `tests.handlers.test_federation.PartialJoinTestCase.test_failed_partial_join_is_clean`
```
2024-08-22 18:57:22-0500 [-] synapse.metrics.background_process_metrics - 253 - ERROR - sync_partial_state_room-0 - Background process 'sync_partial_state_room' threw an exception
	Traceback (most recent call last):
	  File "synapse/synapse/metrics/background_process_metrics.py", line 251, in run
	    return await func(*args, **kwargs)
	           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
	  File "synapse/synapse/handlers/federation.py", line 1842, in _sync_partial_state_room_wrapper
	    await self._sync_partial_state_room(
	  File "synapse/synapse/handlers/federation.py", line 1933, in _sync_partial_state_room
	    await self.state_handler.update_current_state(room_id)
	  File "synapse/synapse/state/__init__.py", line 554, in update_current_state
	    await self._storage_controllers.persistence.update_current_state(room_id)
	  File "synapse/synapse/storage/controllers/persist_events.py", line 491, in update_current_state
	    await self._event_persist_queue.add_to_queue(
	  File "synapse/synapse/storage/controllers/persist_events.py", line 245, in add_to_queue
	    res = await make_deferred_yieldable(end_item.deferred.observe())
	          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	  File "synapse/synapse/storage/controllers/persist_events.py", line 288, in handle_queue_loop
	    ret = await self._per_item_callback(room_id, item.task)
	          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	  File "synapse/synapse/storage/controllers/persist_events.py", line 370, in _process_event_persist_queue_task
	    await self._update_current_state(room_id, task)
	  File "synapse/synapse/storage/controllers/persist_events.py", line 507, in _update_current_state
	    await self.persist_events_store._calculate_sliding_sync_table_changes(
	  File "synapse/synapse/storage/databases/main/events.py", line 624, in _calculate_sliding_sync_table_changes
	    assert most_recent_event_pos_results, (
	           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
	AssertionError: We should not be seeing `None` here because we are still in the room (!room:example.com) and it should at-least have a join membership event that's keeping us here.
```

MadLittleMods committed:
Use `simple_upsert_txn` for `sliding_sync_joined_rooms` in background update

MadLittleMods committed:
Use `simple_upsert_txn` for `sliding_sync_membership_snapshots` in background update

erikjohnston commented on synapse/storage/databases/main/events.py:
```diff
@@ -1956,6 +2806,55 @@ def _store_room_members_txn(
+                    #
+                    # FIXME: We should use `assert_never` here but for some reason
+                    # the exhaustive matching doesn't recognize the `Never` here.
+                    # assert_never(event.membership)
```
I think this is because `Membership` isn't an enum

MadLittleMods replied:
It should work with `Literal` (ideally, it can). It properly narrows the right types in VSCode and comes to `Never` at this spot but I think it's using Pylance instead of mypy. Probably need to wait for feature parity (https://github.com/python/mypy/issues/12535).

We could also do some niceness in `_get_sliding_sync_insert_values_from_state_map(...)` if we had something like https://github.com/python/mypy/issues/16818 to ensure that we're processing every key in the `TypedDict` (`SlidingSyncStateInsertValues`).

MadLittleMods commented on synapse/storage/database.py:
```diff
@@ -1254,9 +1255,9 @@ def simple_upsert_txn(
-        keyvalues: Dict[str, Any],
-        values: Dict[str, Any],
-        insertion_values: Optional[Dict[str, Any]] = None,
+        keyvalues: Mapping[str, Any],
```
To allow for `TypedDict` to be passed here

MadLittleMods committed:
Update comment

MadLittleMods committed:
Split test cases

MadLittleMods committed:
Clear out-of-date rows

MadLittleMods committed:
Add more context for why

MadLittleMods committed:
Why it matters

MadLittleMods committed:
Kick-off background update for out-of-date snapshots

MadLittleMods committed:
Adapt `sliding_sync_joined_rooms` background update to use `event_stream_ordering` for progress

This way we can re-use it for the catch-up background process

MadLittleMods committed:
Note down caveat about `forgotten`

MadLittleMods committed:
Add placeholder tests

MadLittleMods committed:
Fix query in Postgres

MadLittleMods committed:
Maybe: always start background update

MadLittleMods committed:
Start catch-up if nothing written yet

MadLittleMods committed:
Add test for catch-up background update

MadLittleMods committed:
Add test when no rooms

MadLittleMods committed:
Add more tests

MadLittleMods committed:
More tests

MadLittleMods committed:
Round out tests

MadLittleMods committed:
Remove debug logs

MadLittleMods committed:
Use `stream_id` of some point before we fetch the current state

This is simpler and some rooms are so old that they don't have
`current_state_delta_stream` yet. It's easier if we just get a
general max `stream_id` of the whole table than the max `stream_id`
for the specific room anyway.

Thanks @erikjohnston

MadLittleMods committed:
`get_events(...)` will omit events from unknown room versions

Thanks @erikjohnston

MadLittleMods committed:
Add `sliding_sync_joined_rooms_to_recalculate` table

MadLittleMods committed:
Adapt to using `sliding_sync_joined_rooms_to_recalculate` table

MadLittleMods committed:
Add unique index right away for `sliding_sync_joined_rooms_to_recalculate`

This makes it so we can always `upsert` to avoid duplicates otherwise
I'm not sure of how to not insert duplicates in certain situations
(see FIXME in the diff) which would cause problems down the line
for the unique index being added later.

MadLittleMods commented on synapse/storage/schema/main/delta/87/01_sliding_sync_memberships.sql:
```diff
@@ -0,0 +1,153 @@
+-- state events there are for each room.
+CREATE TABLE IF NOT EXISTS sliding_sync_joined_rooms_to_recalculate(
+    room_id TEXT NOT NULL REFERENCES rooms(room_id),
+    PRIMARY KEY (room_id)
```
@erikjohnston I know you mentioned adding the index after the one-shot bulk insert to save on performance when we copy the `rooms` table over but this leads to `unique` problems. We can't upsert without the `unique` index and if we can't upsert, that means we are potentially inserting duplicates which would make the `unique` index fail to be added later.

I've opted to just add the primary index right away. Hopefully since `rooms` also has a primary index on `room_id`, this can be magically smart and fast as well. :grey_exclamation:  Probably would be good to sanity check on your database.

See https://github.com/element-hq/synapse/pull/17512/commits/da463fb102709e12caa3abb5ce2ad3d41bf94401 where I used to add the index in a separate background update and the `FIXME` comment there points out one flaw.

If there is a severe performance penalty, I think we could work around this by making sure we delete any new `room_rows` from `sliding_sync_joined_rooms_to_recalculate` before we insert them. But I'd prefer keeping the index on the table from the beginning as it makes things easier to reason about in general.

erikjohnston replied:
It takes about ~60s to create the table from cold, which I think is fine.

MadLittleMods commented on synapse/storage/databases/main/events.py:
```diff
@@ -261,6 +365,296 @@ async def _persist_events_and_state_updates(
                     (room_id,), frozenset(new_forward_extremities)
                 )
 
+    async def _calculate_sliding_sync_table_changes(
```
TODO: Need to insert rows fully if they don't exist in the table yet

MadLittleMods replied:
Added in https://github.com/element-hq/synapse/pull/17512/commits/bcc3e50a125e37b0f9301913f44bb73191f78488

MadLittleMods commented on synapse/storage/prepare_database.py:
```diff
@@ -567,6 +576,256 @@ def _upgrade_existing_database(
+                key_names=("room_id",),
+                key_values=chunk,
+                value_names=(),
+                # No value columns, therefore make a blank list so that the following
+                # zip() works correctly.
+                value_values=[() for x in range(len(chunk))],
```
The reason I'm using `simple_upsert_many_txn_native_upsert` and `simple_upsert_txn_native_upsert` directly, is because `simple_upsert_many_txn` is not a static method and checks `self._unsafe_to_upsert_tables`. 

It does suck to have to use these workarounds that would be covered if we were using `simple_upsert_many_txn` directly though.

This could be refactored to accommodate a static method version but it doesn't come out very clean for upstream users.

erikjohnston committed:
Remove all rooms pulled out from the queue

erikjohnston committed:
Only run the sliding sync background updates on the main database

erikjohnston committed:
Handle old rows with null event_stream_ordering column

erikjohnston committed:
Fix port script tests by handling empty DBs correctly

erikjohnston committed:
Use event_auth table to get previous membership

MadLittleMods committed:
Clarify why we return early

MadLittleMods committed:
Note why it's there

MadLittleMods committed:
Update metric with how long it took with the index and upsert

MadLittleMods committed:
Ignore membership where we don't have a row in `rooms` table

MadLittleMods committed:
Fix-up `rooms` table for old out-of-band memberships

MadLittleMods committed:
Fill in placeholder issue

MadLittleMods committed:
Fix join condition not working in Postgres

```
common column name "room_id" appears more than once in left table
```

MadLittleMods committed:
Add `instance_name` to `sliding_sync_membership_snapshots`

So we can craft `PersistedEventPosition(...)`

MadLittleMods committed:
Fully-insert `sliding_sync_joined_rooms` rows

This way if the row exists, we can rely on the information in it.
And only use a fallback for rows that don't exist.

MadLittleMods committed:
Merge branch 'develop' into madlittlemods/sliding-sync-pre-populate-room-meta-data

Conflicts:
	synapse/handlers/sliding_sync/__init__.py

MadLittleMods committed:
Explain more in schema

erikjohnston commented:
Woo :tada: :rocket: :tada: :partying_face: :balloon:

MadLittleMods commented:
@erikjohnston Super grateful that you were able to try this PR out multiple times on your own server to find all of the weird old corners in the data that we still need to deal with today :bow:

Thank you @erikjohnston and @reivilibre for all of the discussion so we can ensure that the data is reliable to use before and after the [gradual migration](https://github.com/element-hq/synapse/issues/17623) is fully complete!
